frame/mosaic/src/lib.rs

```diff
@@ -0,0 +1,676 @@
+		/// When we have finality issues occur on the Ethereum chain,
+		/// we burn the locked `IncomingTransaction` for which we know that it is invalid.
+		TransferIntoRescined {
+			account: AccountIdOf<T>,
+			amount: BalanceOf<T>,
+			asset_id: AssetIdOf<T>,
+		},
+		/// The relayer accepted the user's `OutgoingTransaction`.
+		TransferAccepted { from: AccountIdOf<T>, asset_id: AssetIdOf<T>, amount: BalanceOf<T> },
+		/// The user claims his `IncomingTransaction` and unlocks the locked amount.
+		TransferClaimed {
+			by: AccountIdOf<T>,
+			to: AccountIdOf<T>,
+			asset_id: AssetIdOf<T>,
+			amount: BalanceOf<T>,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		BadTTL,
+		BadTimelockPeriod,
+		UnsupportedAsset,
+		NetworkDisabled,
+		UnsupportedNetwork,
+		Overflow,
+		NoStaleTransactions,
+		InsufficientBudget,
+		ExceedsMaxTransferSize,
+		NoClaimableTx,
+		TxStillLocked,
+		NoOutgoingTx,
+		AmountMismatch,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Sets the current relayer configuration. This is enacted immediately and invalidates
+		/// inflight, incoming transactions from the previous relayer. Budgets remain in place
+		/// however.
+		#[pallet::weight(10_000)]
+		pub fn set_relayer(
+			origin: OriginFor<T>,
+			relayer: T::AccountId,
+		) -> DispatchResultWithPostInfo {
+			ensure_root(origin)?;
+			Relayer::<T>::set(StaleRelayer::new(relayer.clone()));
+			Self::deposit_event(Event::RelayerSet { relayer });
+			Ok(().into())
+		}
+
+		/// Rotates the Relayer Account
+		///
+		/// # Restrictions
+		///  - Only callable by the current relayer.
+		///  - TTL must be sufficiently long.
+		#[pallet::weight(10_000)]
+		pub fn rotate_relayer(
+			origin: OriginFor<T>,
+			new: T::AccountId,
+			ttl: T::BlockNumber,
+		) -> DispatchResultWithPostInfo {
+			ensure!(ttl > T::MinimumTTL::get(), Error::<T>::BadTTL);
+			let (relayer, current_block) = ensure_relayer::<T>(origin)?;
+			let ttl = current_block.saturating_add(ttl);
+			let relayer = relayer.rotate(new.clone(), ttl);
+			Relayer::<T>::set(relayer.into());
+			Self::deposit_event(Event::RelayerRotated { account_id: new, ttl });
+			Ok(().into())
+		}
+
+		/// Sets supported networks and maximum transaction sizes accepted by the relayer.
+		#[pallet::weight(10_000)]
+		pub fn set_network(
+			origin: OriginFor<T>,
+			network_id: NetworkIdOf<T>,
+			network_info: NetworkInfo<BalanceOf<T>>,
+		) -> DispatchResultWithPostInfo {
+			ensure_relayer::<T>(origin)?;
+			NetworkInfos::<T>::insert(network_id.clone(), network_info.clone());
+			Self::deposit_event(Event::NetworksUpdated { network_id, network_info });
+			Ok(().into())
+		}
+
+		/// Sets the relayer budget for _incoming_ transactions for specific assets. Does not reset
+		/// the current `penalty`.
+		///
+		/// # Restrictions
+		/// - Only callable by root
+		#[pallet::weight(10_000)]
+		#[transactional]
+		pub fn set_budget(
+			origin: OriginFor<T>,
+			asset_id: AssetIdOf<T>,
```

Comment:
could split assets which are XCMP assets and not? for XCMP assets we do not need transfers - must forbid to mint them.
what do you think?

Reply:
this will ensure we do not hack other parachain

Reply:
This is done by root